pkg/pillar/hypervisor/kvm.go

```diff
@@ -1068,40 +1068,295 @@ func getFmlCustomResolution(status *types.DomainStatus, globalConfig *types.Conf
 	return "", fmt.Errorf("invalid fml resolution %s", fmlResolutions)
 }
 
+type virtualNetwork struct {
+	types.VifConfig
+	networkID   int
+	pciDeviceID int
+}
+
+type pciAddressAllocator struct {
+	pciAssignments           []pciDevice
+	virtualNetworks          []virtualNetwork
+	multifunctionDevices     multifunctionDevs
+	firstFreePCIID           int
+	enforceNetInterfaceOrder bool
+}
+
+// allocate sets pciDeviceID and pciBridgeID for every pciDevice and virtualNetwork
+// based on user-configured ordering requirements.
+func (a *pciAddressAllocator) allocate() error {
```

Comment:
I think I understand how it works... It looks to be correct. But it's definitely hard to verify just by reading the code. I hope the tests cover the complexity of all the cases)

Reply:
I covered all the distinct cases that I could think of in the unit test.